### DISCUSSION TITLE
PopupService to Close Popup

### DISCUSSION THREAD

**Discussion by sonnythao** [CommunityToolkit/Maui#1529] [Q&A] [answered]
@bijington I'm thrilled that the new PopupService is in and I've successfully implemented it in my view model to show a popup - but can I use it to close the opened popup in my view model too?  Before using this new PopupService, I was opening and closing my loading indicator popup from the popup itself as so:

// Open loading indicator
var loadingIndicatorPopup = new LoadingIndicatorPopup
{
    CanBeDismissedByTappingOutsideOfPopup = false
};
Shell.Current.CurrentPage.ShowPopup(loadingIndicatorPopup);
. . . . (other code to run while the popup is opened) . . . .

// Close loading indicator
await loadingIndicatorPopup.CloseAsync();

I opted to use the new PopupService to allow me to more easily write unit tests for my view model.

**Comment by bijington** [comment, marked as answer]
Currently this is not possible, however I believe @VladislavAntonyuk did also suggest we should provide this functionality. I shall write up a proposal to cover this functionality

**Reply by bijington** [reply]
Here is the proposal: https://github.com/CommunityToolkit/Maui/issues/1530

**Reply by ekeneduru** [reply]
Have this been implemented

**Reply by mads195** [reply]
The documentation and samples (links below) suggests yes but on `8.4.0`, no - it doesn't seem that it really is.

This is the sample for `UpdatingPopupViewModel.cs`. As you can see in the OnFinish, there is a popupService.ClosePopup() but we cannot replicate this - ClosePopup() doesn't actually seem to exist. I hope we're wrong!!

```
using CommunityToolkit.Maui.Core;
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;

namespace CommunityToolkit.Maui.Sample.ViewModels.Views;

public partial class UpdatingPopupViewModel(IPopupService popupService) : BaseViewModel
{
	const double finalUpdateProgressValue = 1;
	readonly IPopupService popupService = popupService;

	int updates;

	[ObservableProperty]
	public partial string Message { get; set; } = "";

	[ObservableProperty]
	[NotifyCanExecuteChangedFor(nameof(FinishCommand))]
	public partial double UpdateProgress { get; set; }

	internal async void PerformUpdates(int numberOfUpdates)
	{
		double updateTotalForPercentage = numberOfUpdates + 1;
		updates = numberOfUpdates;

		for (var update = 1; update <= numberOfUpdates; update++)
		{
			Message = $"Updating {update} of {numberOfUpdates}";

			UpdateProgress = update / updateTotalForPercentage;

			await Task.Delay(TimeSpan.FromSeconds(1));
		}

		UpdateProgress = finalUpdateProgressValue;
		Message = $"Completed {numberOfUpdates} updates";
	}

	[RelayCommand(CanExecute = nameof(CanFinish))]
	void OnFinish()
	{
		popupService.ClosePopup();
	}

	[RelayCommand]
	void OnMore()
	{
		popupService.ShowPopup<UpdatingPopupViewModel>(onPresenting: viewModel => viewModel.PerformUpdates(updates + 2));
	}

	bool CanFinish() => UpdateProgress is finalUpdateProgressValue;
}
```


https://learn.microsoft.com/en-us/dotnet/communitytoolkit/maui/views/popup-service#closing-a-popup

**Reply by bijington** [reply]
You are correct, this feature shipped in version [`9.1.1`](https://github.com/CommunityToolkit/Maui/releases/tag/9.1.1) therefore you will need to upgrade the version of the toolkit.
